### PR TITLE
Eng 6511 mview minmax refactor

### PR DESCRIPTION
**Pull request by yzhang1991**

### CODE REVIEW

File: src/frontend/org/voltdb/compiler/DDLCompiler.java [hunk 1]
```diff
@@ -2182,6 +2182,8 @@ void processMaterializedViews(Database db) throws VoltCompiler.VoltCompilerExcep
         }
     }
 
+    private enum MatViewIndexMacthingGroupby {GB_COL_IDX_COL, GB_COL_IDX_EXP,  GB_EXP_IDX_EXP}
```
Comment (paulmartel) [inline src/frontend/org/voltdb/compiler/DDLCompiler.java]: Spelling: Macthing

File: src/frontend/org/voltdb/compiler/DDLCompiler.java [hunk 2]
```diff
@@ -2364,23 +2322,119 @@ private static Index findBestMatchIndexForMatviewMinOrMax(MaterializedViewInfo m
...
+            assert(false);
+            // invalid option
+            return false;
+        }
+
+        // group-by columns/expressions are matched with the corresponding index
+        return true;
+    }
+
+    private static boolean isValidIndexCandiateForMinMax(int idxSize, int gbSize, int diffAllowance) {
```
Comment (paulmartel) [inline src/frontend/org/voltdb/compiler/DDLCompiler.java]: Spelling: Candiate

File: src/frontend/org/voltdb/compiler/DDLCompiler.java [hunk 3]
```diff
@@ -2200,16 +2202,15 @@ private static Index findBestMatchIndexForMatviewMinOrMax(MaterializedViewInfo m
...
+        // return immediately.
+        // If the index found covers only group-by columns (sub-optimal), we will first cache it here.
         Index candidate = null;
         for (Index index : allIndexes) {
-            // matchedAll == true if the index covered all group-by columns (sub-optimal candidate).
-            boolean matchedAll = true;
-            // optimal == true if the index covered both the group-by columns and the min/max aggExpr.
-            boolean optimal = false;
+            // isIndexOptimalForMinMax == true if the index covered both the group-by columns and the min/max aggExpr.
+            boolean isIndexOptimalForMinMax = false;
```
Comment (paulmartel) [inline src/frontend/org/voltdb/compiler/DDLCompiler.java]: Avoid using the exact same name for local variables and methods.

File: src/frontend/org/voltdb/compiler/DDLCompiler.java (hunk 2 above)
Comment (paulmartel) [inline src/frontend/org/voltdb/compiler/DDLCompiler.java]: Could simplify to "return diffAllowance == 1 && idxSize == gbSize + 1;"

File: src/frontend/org/voltdb/compiler/DDLCompiler.java [hunk 4]
```diff
@@ -2364,23 +2322,119 @@ private static Index findBestMatchIndexForMatviewMinOrMax(MaterializedViewInfo m
...
-            if (matchedAll) {
-                // if the index already covered group by columns and the aggCol/aggExpr,
-                // it is already the best index we can get, return immediately.
-                if ( optimal ) {
-                    return index;
-                }
-                // otherwise wait to see if we can find something better!
-                else {
-                    candidate = index;
+                if (! SubPlanAssembler.isPartialIndexPredicateIsCovered(tableScan, coveringExprs, index, exactMatchCoveringExprs)) {
```
Comment (paulmartel) [inline src/frontend/org/voltdb/compiler/DDLCompiler.java]: Nest this block inside the "if (!encodedPredicate.isEmpty()) {" block.

File: src/frontend/org/voltdb/compiler/DDLCompiler.java [hunk 5]
```diff
@@ -2239,115 +2243,69 @@ private static Index findBestMatchIndexForMatviewMinOrMax(MaterializedViewInfo m
...
                 }
-                if ( singleUniqueMinMaxAggExpr != null &&
-                     indexedExprs.size() == groupbyExprs.size() + diffAllowance ) {
-                    if (indexedExprs.get(indexedExprs.size() - 1).equals(singleUniqueMinMaxAggExpr)) {
-                        optimal = true;
-                    }
-                    else {
+
+                if (isValidIndexCandiateForMinMax(indexedExprs.size(), groupbyExprs.size(), diffAllowance)) {
+                    if(! isIndexOptimalForMinMax(matchingCase, singleUniqueMinMaxAggExpr, null, indexedExprs, null)) {
```
Comment (paulmartel) [inline src/frontend/org/voltdb/compiler/DDLCompiler.java]: Always put a space after 'if'

**Comment by paulmartel** [comment]
The new tests (already on master) contain a number of methods that begin "ENG".
Change them to follow the coding standard (methods start with lower case) by renaming them e.g.
from ENG...Insert to insertENG...

File: src/frontend/org/voltdb/compiler/DDLCompiler.java (hunk 4 above)
Comment (yzhang1991) [inline src/frontend/org/voltdb/compiler/DDLCompiler.java]: This change will result in a junit test failure
http://ci.voltdb.lan/view/Branch-jobs/view/branch-ENG-6511-mview-minmax/job/branch-2-community-junit-U10.04-ENG-6511-mview-minmax/19/label_exp=U10.04,target=junit_other/testReport/org.voltdb.compiler/TestDDLCompiler/testIndexedMinMaxViews/

**Pull request closed without merging**